### ISSUE TITLE
Hard-coded criteria for whether a stack frame is "in-project"

### ISSUE THREAD

**Issue by grantovich** [jarednorman/bugsnag-elixir#73]
#17 added detailed reporting of stack frames to Bugsnag. The platform has a notion of whether a given stack frame is "in-project" (i.e. part of your app's code, vs. part of a library), and uses this to control the grouping of error reports.

`bugsnag-elixir` currently makes this call based on whether the file path starts with the hard-coded strings `lib` or `web`. This makes some assumptions about a project's file structure that may not hold true everywhere. At the very least, we should probably make this configurable.

**Comment by grantovich** [comment]
@shantiii Strange coincidence that we noticed this on the same day 😄

Are you in a good position to investigate changing this? My first instinct is that we should reverse the current logic, so a stack frame is marked "in-project" by default unless we know for sure it's in a dependency (or part of the language). I'm not sure off-hand how to make this determination, though. Stack frames from e.g. `System.stacktrace/0` all seem to start with `lib/<lib_name>/...` regardless of whether they're in a dependency or my own app. To classify these properly, we'd have to have users configure a list of app names that are "in-project". Is it possible to have an intelligent default here?

**Comment by shantiii** [comment]
I think allowing people to supply a classifier predicate on bugsnag startup is maybe the easiest way to go: something like
```elixir
in_my_project = &String.contains?(&1, "/my_project_name")
Bugsnag.start(:normal, classifier: in_my_project)
```
and then invoking that predicate on the file names when performing the reporting.
Given the default elixir layout of `lib/project.ex` and `lib/project/....ex`, I think the function specified above is a pretty good default value. What are your thoughts?

**Comment by shantiii** [comment]
Oh, I have some time this weekend I can spend working on this.

**Comment by grantovich** [comment]
Thanks for volunteering 😄 Since explicitly calling `Bugsnag.start` is not required or recommended since 1.4.0 (#35), I think it makes more sense for it to be an app config option like the others.

Since in your example we're working only with the string file path, perhaps being able to provide either a string or a regex would be less friction and cover most common use cases? Example:

```elixir
# uses String.contains?
config :bugsnag, in_project_pattern: "lib/my_project"

# uses String.match?
config :bugsnag, in_project_pattern: ~r/my_project|other_project/
```

It would be impossible to configure a regex via environment variable, but maybe that's fine.

**Comment by shantiii** [comment]
Hey, I didn't end up getting around to it this weekend as I've been on vacay, but there's been internal talk this week about addressing this. Will post an update shortly.

**Comment by grantovich** [comment]
Resolved by #76
